lib/topic_list_responder.rb

```diff
@@ -5,16 +5,23 @@ module TopicListResponder
   def respond_with_list(list)
     discourse_expires_in 1.minute
 
-    respond_to do |format|
-      format.html do
-        @list = list
-        store_preloaded(
-          list.preload_key,
-          MultiJson.dump(TopicListSerializer.new(list, scope: guardian)),
-        )
-        render "list/list"
+    if guardian.anonymous? && SiteSetting.login_required
+      respond_to do |format|
+        format.html { render "default/empty" }
+        format.json {}
+      end
+    else
```

Comment:
Might be "over engineering" but should we do this test _before_ computing the `list`? Since we're not showing anything when login is required, we might as well skip creating a `TopicList` altogether?

Reply:
Feels like we should be doing this test before, maybe in a "before_action" callback?

Reply:
Yes, good call. This was the initial approach, but you're right, it was too late in the cascade of events. 

I have added it now as a `before_action` to the categories, list and topics controllers. Note that this is now a belts and suspenders approach, the preloaded changes also apply here.